File: src/utils/response.ts
import { Response } from 'express';
import { ApiResponse, ApiError } from '../types';
import { AppError } from './errors';

export class ResponseUtils {
  static success<T>(res: Response, data?: T, message?: string): Response<ApiResponse<T>> {
    return res.status(200).json({
      code: 200,
      data,
      message,
    });
  }

  static created<T>(res: Response, data?: T, message?: string): Response<ApiResponse<T>> {
    return res.status(201).json({
      code: 201,
      data,
      message,
    });
  }

  static noContent(res: Response): Response<ApiResponse<void>> {
    return res.status(204).send();
  }

  static error(res: Response, error: AppError | Error): Response<ApiResponse<void>> {
    if (error instanceof AppError) {
      return res.status(error.statusCode).json({
        code: error.statusCode,
        message: error.message,
        errors: error.details ? this.formatErrors(error.details) : undefined,
      });
    }

    return res.status(500).json({
      code: 500,
      message: 'Internal Server Error',
    });
  }

  static badRequest(res: Response, message: string, details?: any): Response<ApiResponse<void>> {
    return res.status(400).json({
      code: 400,
      message,
      errors: details ? this.formatErrors(details) : undefined,
    });
  }

  static unauthorized(res: Response, message: string = 'Unauthorized'): Response<ApiResponse<void>> {
    return res.status(401).json({
      code: 401,
      message,
    });
  }

  static forbidden(res: Response, message: string = 'Forbidden'): Response<ApiResponse<void>> {
    return res.status(403).json({
      code: 403,
      message,
    });
  }

  static notFound(res: Response, message: string = 'Resource not found'): Response<ApiResponse<void>> {
    return res.status(404).json({
      code: 404,
      message,
    });
  }

  static conflict(res: Response, message: string, details?: any): Response<ApiResponse<void>> {
    return res.status(409).json({
      code: 409,
      message,
      errors: details ? this.formatErrors(details) : undefined,
    });
  }

  static paginated<T>(
    res: Response,
    items: T[],
    total: number,
    page: number,
    pageSize: number
  ): Response<ApiResponse<{ items: T[]; total: number; page: number; pageSize: number; totalPages: number }>> {
    const totalPages = Math.ceil(total / pageSize);
    return res.status(200).json({
      code: 200,
      data: {
        items,
        total,
        page,
        pageSize,
        totalPages,
      },
    });
  }

  private static formatErrors(details: any): ApiError[] {
    if (Array.isArray(details)) {
      return details.map((detail: any) => ({
        field: detail.path?.join('.') || detail.field,
        message: detail.message,
        code: detail.code,
      }));
    }

    if (typeof details === 'object' && details !== null) {
      return [{
        field: details.field,
        message: details.message,
        code: details.code,
      }];
    }

    return [{ message: String(details) }];
  }
}

export default ResponseUtils;

File: pkg/utils/response.go
package utils

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"session133/pkg/errors"
)

type Response struct {
	Code    int         `json:"code"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

type PaginatedResponse struct {
	Code      int         `json:"code"`
	Message   string      `json:"message"`
	Data      interface{} `json:"data"`
	Total     int64       `json:"total"`
	Page      int         `json:"page"`
	PageSize  int         `json:"page_size"`
	TotalPages int        `json:"total_pages"`
}

func Success(c *gin.Context, data interface{}) {
	c.JSON(http.StatusOK, Response{
		Code:    200,
		Message: "success",
		Data:    data,
	})
}

func SuccessCreated(c *gin.Context, data interface{}) {
	c.JSON(http.StatusCreated, Response{
		Code:    201,
		Message: "created",
		Data:    data,
	})
}

func SuccessPaginated(c *gin.Context, data interface{}, total int64, page, pageSize int) {
	totalPages := int(total) / pageSize
	if int(total)%pageSize > 0 {
		totalPages++
	}
	c.JSON(http.StatusOK, PaginatedResponse{
		Code:       200,
		Message:    "success",
		Data:       data,
		Total:      total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: totalPages,
	})
}

func Error(c *gin.Context, err *errors.AppError) {
	httpStatus := http.StatusInternalServerError
	switch err.Code {
	case errors.ErrCodeInvalidParams:
		httpStatus = http.StatusBadRequest
	case errors.ErrCodeUnauthorized:
		httpStatus = http.StatusUnauthorized
	case errors.ErrCodeForbidden:
		httpStatus = http.StatusForbidden
	case errors.ErrCodeNotFound:
		httpStatus = http.StatusNotFound
	case errors.ErrCodeConflict:
		httpStatus = http.StatusConflict
	case errors.ErrCodeRateLimitExceeded:
		httpStatus = http.StatusTooManyRequests
	case errors.ErrCodeServiceUnavailable:
		httpStatus = http.StatusServiceUnavailable
	}
	c.JSON(httpStatus, Response{
		Code:    int(err.Code),
		Message: err.Message,
		Data:    gin.H{"details": err.Details},
	})
}
